### PR TITLE
feat(model): add Data preparation for lightning fabric

### LINKED ISSUE #157: [Super AI] Prepare Lightning Fabric Pipeline for continue pretraining in Thai language

### PR DESCRIPTION
**Pull request by MoosaTae**
## Why this PR
lightning fabric need data preparation for faster iteration

## Changes
- rename old datasetWrapper
- add TokenizedDataset and TokenDatasetWrapper class
- add parser argument  streaming, dataset_name_or_path and dataset_dir
- add new tutorial
## preprocess dataset if don't want to use streaming dataset
```python src/model/scripts/lighting_training/data_preprocessing.py --mode train --model path/to/tokenizer --max_tokens 2048 --save_path ./tokendata```
argument of preprocessing
- mode: train | val
- model
- max_tokens
- save_path
- chunk_size
- batch_size
- num_proc
- dataset_name
- dataset_dir
### test process oscar unshuffled_deduplicated_th 1.0.0 
on 128 CPUs max tokens 2048 chunk_size 1024^2 batch_size 25000
image

## Related Issues
Close #157 

## Checklist
- [x] PR should be in the [Naming convention](../../docs/PR_NAMING.md)
- [x] Assign yourself in to Assigneees
- [x] Tag related issues
- [x] Constants name should be ALL_CAPITAL, function name should be snake_case, and class name should be CamelCase
- [x] complex function/algorithm should have [Docstring](https://peps.python.org/pep-0257/)
- [x] 1 PR should not have more than 200 lines changes (Exception for test files). If more than that please open multiple PRs
- [x] At least PR reviewer must come from the task's team (model, eval, data)

### CODE REVIEW

**Review by boat1603 — APPROVED** [review]

File: src/model/scripts/lighting_training/data_preprocessing.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--mode",
+        type=str,
+        default="train",
+        help="train | val",
+    )
+    parser.add_argument(
```
Comment (new5558) [inline src/model/scripts/lighting_training/data_preprocessing.py:14]: Naming tokenizer name as ‘model’ can cause confusion?

File: src/model/scripts/lighting_training/data_preprocessing.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        "--batch_size",
+        type=int,
+        default=25000,
+    )
+    parser.add_argument(
+        "--num_proc",
+        type=int,
+        default=128,
+    )
+    parser.add_argument(
```
Comment (new5558) [inline src/model/scripts/lighting_training/data_preprocessing.py:45]: Our input dataset may not be huggingface dataset but multiple jsonl files. How can we handle this
Options I can think of
1. Create simple script to convert jsonl to huggingface dataset and also create simple script to merge multiple hf dataset into one hf dataset. Then we use one hf dataset into this tokenization pipeline.
2. Modify this pipeline to also support multiple jsonl and converted into one hf dataset after tokenized

**Comment by new5558** [comment]
Todo (me): refactor cli to typer

**Comment by MoosaTae** [comment]
what is cli? @new5558

**Comment by new5558** [comment]
> what is cli? @new5558

Command line. Never, mind @boat1603 will refactor script himself

**Review by boat1603 — COMMENTED** [review]
Please resolve conflict

File: src/model/scripts/lighting_training/data_preprocessing.py (hunk 1 above)
Comment (MoosaTae) [reply src/model/scripts/lighting_training/data_preprocessing.py:14]: done

File: src/model/scripts/lighting_training/data_preprocessing.py (hunk 2 above)
Comment (MoosaTae) [reply src/model/scripts/lighting_training/data_preprocessing.py:45]: done

File: src/model/openthaigpt_pretraining_model/data_wrapper.py [hunk 3]
```diff
@@ -1,37 +1,176 @@
...
+            self.data_set = load_dataset(
+                dataset_name,
+                dataset_dir,
+                split=SPLIT_TRAIN,
+            )
+        else:
+            raise NotImplementedError("only support Train,Val")
+        self.num_shards = (len(self.data_set) + chunk_size - 1) // chunk_size
 
-class DatasetWrapper(IterableDataset):
```
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/data_wrapper.py:7]: Why you remove this? We are using this class [here](https://github.com/OpenThaiGPT/openthaigpt-pretraining/blob/1e1c481f78be18552bb154fdf66d78e761560817/src/model/openthaigpt_pretraining_model/optimizers/lion/utils.py#L25).

File: src/model/openthaigpt_pretraining_model/data_wrapper.py [hunk 4]
```diff
@@ -1,37 +1,176 @@
...
+        tokenizer: str,
+        max_tokens: int = 2048,
+        save_path: str = "./",
+        chunk_size: int = 1024 * 1024,
+        batch_size: int = 10000,
+        num_proc: int = 16,
+        dataset_name: str = "oscar",
+        dataset_dir: str = "unshuffled_deduplicated_th",
+    ):
+        if len(findall("llama", tokenizer)):
```
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/data_wrapper.py:27]: We should load hf dataset before using `TokenizedDataset`. We have an example at [here](https://github.com/OpenThaiGPT/openthaigpt-pretraining/blob/1e1c481f78be18552bb154fdf66d78e761560817/src/model/openthaigpt_pretraining_model/lightning/utils.py#L92).

File: src/model/openthaigpt_pretraining_model/data_wrapper.py [hunk 5]
```diff
@@ -1,37 +1,176 @@
...
+        else:
+            self.tokenizer = AutoTokenizer.from_pretrained(tokenizer)
+
+        self.mode = mode
+        self.max_tokens = max_tokens
+        self.save_path = save_path
+        self.chunk_size = chunk_size
+        self.batch_size = batch_size
+        self.num_proc = num_proc
+        if mode == "val":
```
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/data_wrapper.py:40]: You can use `data_set` as class input.

File: src/model/openthaigpt_pretraining_model/data_wrapper.py [hunk 6]
```diff
@@ -1,37 +1,176 @@
...
-                structure.
-            text_column_name: Column name which contain text data.
-        """
-        self.tokenizer = tokenizer
-        self.dataset = dataset
+        self.mode = mode
         self.max_tokens = max_tokens
-        self.text_column_name = text_column_name
+
+        if len(findall("llama", model_or_path)):
```
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/data_wrapper.py:94]: Same above

File: src/model/openthaigpt_pretraining_model/data_wrapper.py [hunk 7]
```diff
@@ -1,37 +1,176 @@
...
+            self.tokenizer = LlamaTokenizer.from_pretrained(model_or_path)
+        else:
+            self.tokenizer = AutoTokenizer.from_pretrained(model_or_path)
+
+        if mode == "val":
```
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/data_wrapper.py:99]: Same above

File: src/model/openthaigpt_pretraining_model/data_wrapper.py [hunk 8]
```diff
@@ -1,37 +1,176 @@
...
+            )
+        else:
+            raise NotImplementedError("only support Train,Val")
+        self.num_shards = (len(self.data_set) + chunk_size - 1) // chunk_size
 
-class DatasetWrapper(IterableDataset):
+        self.tokenized_data = None
+
+    def tokenize_data(self):
+        def tokenize_function(examples):
```
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/data_wrapper.py:59]: You can write
```python
class TokenizedDataset:
    ...
    def tokenize_function(self, examples): # Please use variable better than `examples`
        """
        Description:
            ...
        Args:
            ...
        """
        tokenized_text = self.tokenizer(
            examples["text"],
            truncation=True,
            padding=True,
            max_length=self.max_tokens,
        )["input_ids"]
        return {"tokenized_text": tokenized_text}
    def tokenize_data(self):
        ...
        tokenized_dataset = chunk.map(
            self.tokenize_function,
            batched=True,
            batch_size=self.batch_size,
            num_proc=self.num_proc,
        )
```

File: src/model/openthaigpt_pretraining_model/data_wrapper.py [hunk 9]
```diff
@@ -1,37 +1,176 @@
...
+            tokenized_text = self.tokenizer(
+                examples["text"],
+                truncation=True,
+                padding=True,
+                max_length=self.max_tokens,
+            )["input_ids"]
+            return {"tokenized_text": tokenized_text}
```
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/data_wrapper.py:66]: Please use CONST

File: src/model/openthaigpt_pretraining_model/lightning/utils.py
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/lightning/utils.py:1]: You should upgrade this with [lightning_trainer.py](https://github.com/OpenThaiGPT/openthaigpt-pretraining/blob/1e1c481f78be18552bb154fdf66d78e761560817/src/model/openthaigpt_pretraining_model/trainer/lightning_trainer.py)

File: src/model/openthaigpt_pretraining_model/lightning/utils.py [hunk 10]
```diff
@@ -120,17 +121,36 @@ def __init__(
             )
         else:
             raise NotImplementedError("only support Llama, llama_hf or GPTJ")
-        train_dataset = get_dataset(dataset_name, split=SPLIT_TRAIN, shuffle=True)
```
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/lightning/utils.py:123]: You should not delete this.

File: src/model/openthaigpt_pretraining_model/data_wrapper.py (hunk 3 above)
Comment (MoosaTae) [reply src/model/openthaigpt_pretraining_model/data_wrapper.py:7]: sorry about that

Comment (MoosaTae) [reply src/model/openthaigpt_pretraining_model/data_wrapper.py]: I restored it now

File: src/model/openthaigpt_pretraining_model/lightning/utils.py
Comment (MoosaTae) [reply src/model/openthaigpt_pretraining_model/lightning/utils.py:1]: what is different between utils in lightning and lightning_trainer that you mentioned?

Comment (boat1603) [reply src/model/openthaigpt_pretraining_model/lightning/utils.py:1]: It’s the refactor one. We will remove this utils.py file after we refactor all training script in repo.

File: src/model/openthaigpt_pretraining_model/data_wrapper.py (hunk 4 above)
Comment (MoosaTae) [reply src/model/openthaigpt_pretraining_model/data_wrapper.py:62]: do you mean input to be datasets and tokenize?

Comment (boat1603) [reply src/model/openthaigpt_pretraining_model/data_wrapper.py:62]: Yes

Comment (boat1603) [reply src/model/openthaigpt_pretraining_model/data_wrapper.py:62]: We may have customized dataset or tokenizer we should handle these once.

File: src/model/openthaigpt_pretraining_model/lightning/utils.py (hunk 10 above)
Comment (MoosaTae) [reply src/model/openthaigpt_pretraining_model/lightning/utils.py:123]: what about streaming dataset?

File: src/model/openthaigpt_pretraining_model/lightning/utils.py
Comment (MoosaTae) [reply src/model/openthaigpt_pretraining_model/lightning/utils.py:1]: should I remove now?

File: src/model/openthaigpt_pretraining_model/data_wrapper.py (hunk 4 above)
Comment (MoosaTae) [reply src/model/openthaigpt_pretraining_model/data_wrapper.py:62]: one is datasets already tokenized and the other is not.

File: src/model/openthaigpt_pretraining_model/lightning/utils.py (hunk 10 above)
Comment (boat1603) [reply src/model/openthaigpt_pretraining_model/lightning/utils.py:123]: We should handle `streaming dataset` in `get_dataset`.

File: src/model/openthaigpt_pretraining_model/data_wrapper.py (hunk 4 above)
Comment (MoosaTae) [reply src/model/openthaigpt_pretraining_model/data_wrapper.py]: done

File: src/model/openthaigpt_pretraining_model/lightning/utils.py
Comment (MoosaTae) [reply src/model/openthaigpt_pretraining_model/lightning/utils.py:1]: okay

**Comment by boat1603** [comment]
Please update your branch

**Review by boat1603 — APPROVED** [review]
This PR I think update your branch and make our code work is OK. For refactoring, we will do this in the next PR.

**Review by new5558 — APPROVED** [review]

**Pull request merged**